### PR TITLE
sql: don't grant privileges to nonexistent users

### PR DESCRIPTION
**Pull request by pananc**
related to #20226

### CODE REVIEW

File: pkg/sql/logictest/testdata/logic_test/grant_database [hunk 1]
```diff
@@ -12,6 +12,15 @@ a        root ALL
...
 REVOKE SELECT ON DATABASE a FROM root
 
+statement ok
+CREATE USER readwrite
+
+statement error pq: user "test-user" does not exist
+GRANT ALL ON DATABASE a TO readwrite, "test-user"
+
+statement ok
+insert into system.users values('test-user','');
```
Comment (mberhault) [inline pkg/sql/logictest/testdata/logic_test/grant_database]: nit: we usually capitalize the SQL keywords.

File: pkg/sql/user.go [hunk 2]
```diff
@@ -55,3 +55,19 @@ func GetUserHashedPassword(
 
 	return exists, hashedPassword, err
 }
+
+// GetAllUsers returns all the usernames in system.users.
+func GetAllUsers(ctx context.Context, plan *planner) (map[string]bool, error) {
+	query := `SELECT username FROM system.users`
+	p := makeInternalPlanner("get-all-user", plan.txn, security.RootUser, plan.session.memMetrics)
+	defer finishInternalPlanner(p)
+	rows, err := p.queryRows(ctx, query)
```
Comment (mberhault) [inline pkg/sql/user.go]: you should return here if `err != nil`

**Review by mberhault — COMMENTED** [review]
Thanks for renaming. You didn't have to create a new PR though, just `git rebase --interactive` to change the commit description and push. The github PR title can be edited as well (it doesn't refresh from the commit description)

File: pkg/server/admin_test.go [hunk 3]
```diff
@@ -437,6 +444,8 @@ func TestAdminAPITableDetails(t *testing.T) {
 							default2 INT DEFAULT 2,
 							string_default STRING DEFAULT 'default_string'
 						)`, escDBName, escTblName),
+				fmt.Sprintf("CREATE User readonly"),
```
Comment (mberhault) [inline pkg/server/admin_test.go]: `USER` should be capitalized.

**Comment by pananc** [comment]
@mberhault   Hello, i have done according to your comments, can you help me review my patch again, thanks.

File: pkg/sql/grant_revoke.go [hunk 4]
```diff
@@ -58,6 +59,18 @@ func (p *planner) changePrivileges(
 	grantees tree.NameList,
 	changePrivilege func(*sqlbase.PrivilegeDescriptor, string),
 ) (planNode, error) {
+	//check whether grantees exists
```
Comment (mberhault) [inline pkg/sql/grant_revoke.go]: Sorry to be so picky, but could you add a space at the beginning of the comment.

File: pkg/sql/grant_revoke.go [hunk 5]
```diff
@@ -58,6 +59,18 @@ func (p *planner) changePrivileges(
...
+	users, err := GetAllUsers(ctx, p)
+
```
Comment (mberhault) [inline pkg/sql/grant_revoke.go]: And remove this newline.

File: pkg/sql/user.go [hunk 6]
```diff
@@ -55,3 +55,20 @@ func GetUserHashedPassword(
...
+	if err != nil {
+		return nil, err
+	}
+
+	users := make(map[string]bool)
+	for _, row := range rows {
+		users[string(tree.MustBeDString(row[0]))] = true
+	}
+	return users, err
```
Comment (mberhault) [inline pkg/sql/user.go]: You can return `users, nil` here.

**Review by mberhault — APPROVED** [review]
LGTM. Sorry, just a few last nitpicks.

**Comment by pananc** [comment]
@mberhault  thank you for reviewing my patch, all comments are helpful. I have recommit my patch, plz help me review again.

**Comment by mberhault** [comment]
LGTM. Thank you for doing this and thank you for your patience. I'll merge once the tests are done running.

**Pull request merged**